### ISSUE TITLE
[BUG] eval_conflict_error: object keys must be unique when evaluating terraform plan file

### ISSUE THREAD

**Issue by james1miller93** [fugue/regula#256]
**Describe the bug**
We're hitting the `eval_conflict_error: object keys must be unique` error when evaluating a large terraform plan file.

```
FATAL lib/fugue/resource_view/terraform.rego:120: eval_conflict_error: object keys must be unique
```

This is a legacy project which we're migrating to a new process which includes regula policy evaluation. The projects state already contains a large number of resources and nested modules.

It looks like a bug in the module parser [here](https://github.com/fugue/regula/blob/961732acedf1a6e792c005d5add19868f818770b/rego/lib/fugue/resource_view/terraform.rego#L120) is causing the error.

We're running the regula library and policies via conftest, but I've also reproduced the error when running the config through the regula binary.

I've also tested this with a plan generated by multiple versions of terraform (including latest) and the behaviour is the same.

**How you're running Regula**
Please include versions of all relevant tools. Some examples:
* Using Regula v2.1.0 as a Rego library with Conftest v0.23.0 and a Terraform plan JSON input that I generated with Terraform v1.0.11 - this just hangs when run locally. This is how we run regula in our pipelines and the error has been produced in this environment.
```shell
terraform plan -out=terraform.tfplan --lock=false
terraform show -json terraform.tfplan > plan.json
conftest pull --policy ./policy "github.com/fugue/regula//rego/conftest?ref=v2.1.0"
conftest pull --policy ./policy/regula/lib "github.com/fugue/regula//rego/lib?ref=v2.1.0"
conftest pull --policy ./policy/regula/rules "github.com/fugue/regula//rego/rules?ref=v2.1.0"
conftest test ./policy plan.json
```

* Using Regula v2.1.0 as a CLI tool and a Terraform plan json file as an input.
```shell
regula run plan.json --verbose
```

**Operating System**
macOS Monterey 12.0.1
x86_64

**IaC Configuration**
Unfortunately, I cannot share the full terraform configuration. However, I may be able to share the output of `terraform state list` if required.

**Additional context**
The state in this project does contain over 1500 resources, so the plan file we're evaluating is large.

**Comment by jaspervdj-luminal** [comment]
@james1miller93 Thanks for the detailed info!  I had a quick try at reproducing this but haven't had luck so far.

We'll keep investigating, in the meantime, these are two things that could be helpful to us to narrow this down further:

1. Does this issue also occur when you run the `regula` binary against the `.tf` files directly (rather than the plan)?  We've made some improvements to this lately, and it could be a temporary workaround (though I understand it wouldn't fit into your workflow easily).

2. The following command should print a highly redacted overview of just the module hierarchy (where I also suspect the bug is).  If this does not contain any sensitive details, would it be possible to post this here or send it to jasper@fugue.co?  Feel free to censor out any further details if necessary:

```bash
jq '.configuration.root_module | del(.outputs) | del(.resources) | walk(if type == "object" and .module.resources then del(.module.resources) | del(.module.outputs) | del(.expressions) | del(.module.variables) | del(.source) else . end)' <plan.json
```

**Comment by james1miller93** [comment]
@jaspervdj-luminal - Thanks for the pointers and for looking into this for me.

1. Running the regula binary directly against the `.tf` files does work. As mentioned above, this doesn't fit into our current workflow, but does give us something to work with in the meantime.

2. I've sent the redacted output to you via email. I've redacted a few additional pieces of data in the file, but the module/object names remain untouched and I've made it clear where the redacted info is missing.

Let me know if you need any further details, or if you need me to run any additional tests.

**Comment by james1miller93** [comment]
@jaspervdj-luminal - Have you had a chance to look into this issue?

**Comment by jaspervdj-luminal** [comment]
@james1miller93 Apologies, priorities got shuffled around a bit and then the holidays hit.  I will be looking into this today.

**Comment by jaspervdj-luminal** [comment]
@james1miller93 I think I was able to reproduce the issue using the redacted plan.  Thanks a lot for providing that!  If this is at all possible, could you give the [fix/RM-6363/module-ambiguity](https://github.com/fugue/regula/tree/fix/RM-6363/module-ambiguity) branch a try?  I believe it could fix the issue.

**Comment by jaspervdj-luminal** [comment]
@james1miller93 I'll add a test and get this merged into master, that way it should be in the release next Thursday.

**Comment by jaspervdj-luminal** [comment]
We've found an issue in master and will have to delay the release until Monday, my apologies.

**Comment by mishrasanjeev224** [comment]
Hi @jaspervdj-luminal I haven't seen any progress since last couple of weeks. Could you please update and let us know when this issue can be fixed. We are blocked on our migration work due to this issue. Will be appreciated if we prioritise this issue and get this fixed ASAP.

**Comment by jaspervdj-luminal** [comment]
@mishrasanjeev224 Hey, this fix was included in the [2.4.0 release](https://github.com/fugue/regula/releases/tag/v2.4.0).  Please let us know if this works for your case?

**Comment by Saathveegan** [comment]
Hi,
Looks like same error was introduced from version 2.7.0.

level=debug msg="loadedPaths[plan.json] -> plan.json"
level=fatal msg="lib/aws/s3/s3_library.rego:38: eval_conflict_error: object keys must be unique"

Please let me know if you need any more information.

Regards,
Saath
